simphony/cuds/model.py

```diff
@@ -0,0 +1,170 @@
+
+        # Memory storage to keep CUDS components
+        self._store = {}
+
+        # The generic data container
+        self._data = DataContainer()
+
+        # A map to find the object for a given id amoung datasets or
+        # simple components. Unfortunately, at the moment dataset
+        # containers do not have `uid' and therefore this workaround
+        # is necessary. This is a dict of key of any kind and a lambda
+        # that will return the corresponding value.
+        self._map = {}
+
+    @staticmethod
+    def _is_cuds_component(obj):
+        """Check whether the object is a CUDS component."""
+        # When I see a bird that walks like a duck and swims like a duck
+        # and quacks like a duck, I call that bird a duck
+        # TODO: replace with metadata type checks
+        return all([hasattr(obj, 'name'),
+                    hasattr(obj, 'data')])
+
+    @staticmethod
+    def _is_dataset(obj):
+        """Check if the object is a dataset."""
+        if isinstance(obj, (ABCParticles, ABCLattice, ABCMesh)):
+            return True
+        else:
+            return False
+
+    @property
+    def data(self):
+        """Data container for CUDS items."""
+        return DataContainer(self._data)
+
+    @data.setter
+    def data_setter(self, value):
+        self._data = DataContainer(value)
+
+    def add(self, component):
+        """Add a component to the CUDS computational model.
+
+        Parameters
+        ----------
+        component: CUDSComponent
+            a component of CUDS, reflecting SimPhoNy metadata
+        """
+        # Check if the object is valid
+        if not self._is_cuds_component(component):
+            raise ValueError('Not a CUDS component.')
+
+        # Add datasets separately
+        if self._is_dataset(component):
+            self._dataset_store.add(component)
+            # Datasets have a uuid field called uid
+            self._map[component.name] = \
+                lambda key=component.name: self._dataset_store.get(key)
+        else:
+            # Store the component. Any cuds item has uuid property.
+            component_id = str(component.uuid)
+            self._store[component_id] = component
+            self._map[component_id] = \
+                lambda key=component_id: self._store.get(key)
+
+    def get(self, component_id):
+        """Gets a component from the CUDS computational model.
+
+        Parameters
+        ----------
+        component_id: uuid.UUID
+            key of a CUDS component.
+
+        Returns
+        -------
+        component: CUDSComponent
+            a cuds component
+        """
+        if not isinstance(component_id, (str, uuid.UUID)):
+            raise TypeError('ID should be of string or UUID type')
+        if isinstance(component_id, uuid.UUID):
+            component_id = str(component_id)
+        if component_id in self._map:
+            return self._map[component_id]()
+
```

Comment:
Good question. Since datasets do not have uid, I had to implement a workaround for now (not to duplicate everything just for them). 

So, probably yes, we will add uid to datasets too.